design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,632 @@
+  _AWS Secret Manager_.
+- **Additional configuration** to reach to the store like its endpoint and
+credentials to authenticate/authorize. For example, a `kubeconfig` for 
+_Kubernetes_, a `server` endpoint + auth config for _Vault_ and access/secret
+keys for _AWS Secret Manager_.
+
+### API
+
+#### Secret Configuration: PublishConnectionDetailsTo
+
+We will deprecate the existing `writeConnectionSecretToRef` field in favor of
+`publishConnectionDetailsTo` field which would support publishing connection
+details to the local Kubernetes cluster **or** to an external secret store.
+
+We will take the **name** of secret and some per secret metadata like tags in
+our secret configuration API (i.e. `publishConnectionDetailsTo`). The rest of
+the configuration will go to a separate store specific config (`StoreConfig`).
+This classification enables building a flexible API that satisfies the
+separation of concerns between platform operators and consumers which Crossplane
+already enables today for Kubernetes Secrets. 
+
+We will end up having a unified configuration spec for all external secret store
+types which contains the name field (`name`) and a reference to any additional
+store specific configuration (`configRef`). This would be enough to uniquely
+identify and access any secret instance, however there could still be some
+additional attributes or metadata specific to store type that might be desired
+to be set per secret instance. For example, `labels`, `annotations` and `type`
+of the secret in _Kubernetes_; `Tags` and `EncryptionKey` in _AWS_.
+
+A `StoreConfig` named `default` will be created during installation which is
+configured to write secret to a Kubernetes cluster in the Crossplane
+installation namespace. 
+
+`StoreConfig.configRef` will be an optional parameters and **in its absence**:
+
+- For cluster-scoped resources (e.g. `MR` and `XR`), it will be late initialized
+as `configRef.name=default`.
+- For namespaced resources (e.g. `XRC`), secrets published to the same
+namespace as the resource.
+
+**Examples:**
+
+Publish Connection details to a _Kubernetes_ secret (already existing case):
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    configRef:
+      name: default
+```
+
+Publish Connection details to a _Vault_ secret:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    configRef:
+      name: vault-dev
+```
+
+Publish Connection details to a _Kubernetes_ secret with some labels/annotations:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    attributes:
+      labels:
+        environment: production
+      annotations:
+        acme.example.io/secret-type: infrastructure
+    configRef:
+      name: default
+```
+
+Publish Connection details to an _AWS Secret Manager_ secret with some tags:
+
+```yaml
+spec:
+  publishConnectionDetailsTo:
+    name: my-db-connection
+    attributes:
+      tags:
+        environment: production
```

Comment:
Hmm, I was always thinking that the author of this spec would be aware of the secret store to be used, but this might not be the case 🤔 But not sure how to support cases like [this](https://github.com/crossplane/crossplane/pull/2494#discussion_r754145633) in a generic way where we need to configure multiple attributes (labels/annotations/types) of the secrets. 
One solution just coming to my mind is, instead of arbitrary key value pairs we stick to Kubernetes ones, but each store config implementation maps them internally.

For example, for the following updated spec:

```
spec:
  publishConnectionDetailsTo:
    name: my-db-connection
    labels:
      environment: production
    annotations:
      tekton.dev/git-0: https://github.com
    type: kubernetes.io/basic-auth
    configRef:
      name: default
```

AWS secret store implementation would use `labels` as `tags` but ignore others (annotations and type).
What do you think?

Reply:
If we agree on this, I think I will also rename `defaultScope` as `defaultNamespace` and it would be up to store config implementation how to map namespace in k8s.

Reply:
Those both sound good to me. I'd suggest moving labels and annotations out from under attribs to the same level as name for maximum symmetry. Or rename attribs to 'metadata'.

Reply:
Renamed `attributes` as `metadata` 👍